disco_aws_automation/disco_deploy.py

```diff
@@ -334,44 +323,56 @@ def run_tests_with_maintenance_mode(self, ami):
             self._set_maintenance_mode(hostclass, self._get_old_instances(ami.id), False)
         return ret
 
-    def handle_tested_ami(self, old_hostclass_dict, ami, desired_size, run_tests=False, dry_run=False):
+    def handle_tested_ami(self, ami, pipeline_dict=None, run_tests=False, dry_run=False, old_group=None):
         '''
         Tests hostclasses which we can deploy normally.
 
         Deploys AMIs inside the same autoscaling group, and destroys old instances on successful passing,
         otherwise destroys the new instances and rolls back the ASG's configuration.
         '''
-        logging.info("testing deployable hostclass %s AMI %s with %s deployment strategy",
-                     old_hostclass_dict["hostclass"], ami.id, DEPLOYMENT_STRATEGY_CLASSIC)
+        hostclass = DiscoBake.ami_hostclass(ami)
+
+        logging.info(
+            "testing deployable hostclass %s AMI %s with %s deployment strategy",
+            hostclass,
+            ami.id,
+            DEPLOYMENT_STRATEGY_CLASSIC
+        )
 
         if dry_run:
             return
 
-        if desired_size and run_tests and not self.run_integration_tests(ami):
+        if not pipeline_dict:
+            pipeline_dict = {}
+
+        if old_group and run_tests and not self.run_integration_tests(ami):
             raise Exception("Failed pre-test -- not testing AMI {}".format(ami.id))
 
-        new_size = desired_size * 2 if desired_size else 1
-        new_hostclass_dict = copy.deepcopy(old_hostclass_dict)
-        new_hostclass_dict["sequence"] = 1
-        new_hostclass_dict["max_size"] = new_size
-        new_hostclass_dict["min_size"] = new_size / 2
-        new_hostclass_dict["desired_size"] = new_size
-        new_hostclass_dict["smoke_test"] = "no"
-        new_hostclass_dict["ami"] = ami.id
-        post_hostclass_dict = copy.deepcopy(new_hostclass_dict)
+        # Generate the two pipelines we'll need, one with double instance sizing for deployment, and another
+        # that has correct instance sizing for the final form of the ASG.
+        new_hostclass_dict = self._generate_deploy_pipeline(
+            deployment_strategy=DEPLOYMENT_STRATEGY_CLASSIC,
+            pipeline_dict=pipeline_dict,
+            old_group=old_group,
+            ami=ami
+        )
+
+        post_hostclass_dict = self._generate_post_deploy_pipeline(
+            pipeline_dict=pipeline_dict,
+            old_group=old_group,
+            ami=ami
+        )
 
         self._disco_aws.spinup([new_hostclass_dict])
 
         try:
-            if (self.wait_for_smoketests(ami.id, desired_size or 1) and
+            if (self.wait_for_smoketests(ami.id, new_hostclass_dict["desired_size"]) and
```

Comment:
This desired size is supposed that of post_hostclass_dict not of new_hostclass_dict